pkg/fleetautoscalers/controller.go

```diff
@@ -293,3 +324,62 @@ func (c *Controller) updateStatusUnableToScale(ctx context.Context, fas *autosca
 
 	return nil
 }
+
+// createFasThread creates a FleetAutoScaler sync routine
+func (c *Controller) createFasThread(ctx context.Context, fas *autoscalingv1.FleetAutoscaler) error {
+	key := fas.Namespace + "/" + fas.Name
+	ticker := time.NewTicker(time.Duration(fas.Spec.Sync.FixedInterval.Seconds) * time.Second)
+	c.fasThreads[key] = fasThread{
+		terminateSignal: make(chan struct{}),
+		generation:      fas.Generation,
+	}
+	// scale immediately when an FAS is created or updated
+	if err := c.fleetAutoScale(ctx, key); err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-c.fasThreads[key].terminateSignal:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				c.loggerForFleetAutoscalerKey(key).Debug("run routine sync")
+				err := c.fleetAutoScale(ctx, key)
+				if err != nil {
+					c.loggerForFleetAutoscaler(fas).WithError(err).Error("routine sync failed")
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// removeFasThread remove a FleetAutoScaler sync routine
+func (c *Controller) removeFasThread(fas *autoscalingv1.FleetAutoscaler) {
+	key := fas.Namespace + "/" + fas.Name
+	if _, ok := c.fasThreads[key]; ok {
+		c.fasThreads[key].terminateSignal <- struct{}{}
+		delete(c.fasThreads, key)
+	}
+}
+
+// syncFleetAutoscalerWithCustomSyncInterval syncs the fleet based on a custom interval configured in Fas
+func (c *Controller) syncFleetAutoscalerWithCustomSyncInterval(ctx context.Context, key string) error {
+	fas, err := c.getFleetAutoscalerByKey(key)
+	if fas == nil {
+		return err
+	}
+	thread, ok := c.fasThreads[key]
+	if !ok {
+		return c.createFasThread(ctx, fas)
+	}
+	if fas.Generation != thread.generation {
```

Comment:
@roberthbailey I'm 97.3% sure this correct to use (i.e. if the generation has changed, we know the spec has changed/it has been redeployed), but want to run it past you in case we got something wrong.

Reply:
Let me make sure I understand the design.

* Each fleet autoscaler gets a separate go routine which is always running in the background. Every `fas.Spec.Sync.FixedInterval.Seconds` it checks to see if the fleet should be scaled.
* Meanwhile, when this sync function is called, it may decide to remove and recreate the go routine for a fleet autoscaler because the generation number has changed. This will cause the fleet to be scaled immediately (not waiting for the next timer tick).

According to https://github.com/kubernetes/kubernetes/issues/67428 the generation for a CRD will only change when the spec changes (but **not** metadata like annotations or labels). 

So if the intent is to look only for spec changes, then this looks ok. But if changing annotations or labels should trigger an immediate scale check, then this is insufficient.

Reply:
That's exactly what I expect. We will ignore changes on annotations or labels.